Adds always on display of media URL

draganescu opened the pull request:
## Description
Implementation of one of the variants in #18992 about always showing the URL of the current media in the media replace control.

## How has this been tested?
Tested locally.

## Screenshots <!-- if applicable -->

![media-replace-url](https://user-images.githubusercontent.com/107534/73757711-4f453180-4772-11ea-8090-242a13aa6b06.gif)



## Types of changes
Updated `MediaReplaceFlow` and `LinkViewer`.
This should eventually use the new `LinkControl` instead.

cc: @shaunandrews

Soean commented on packages/block-editor/src/components/media-replace-flow/index.js:
```diff
@@ -174,15 +173,9 @@ const MediaReplaceFlow = (
 					</NavigableMenu>
-					{ showURLInput && <div className="block-editor-media-flow__url-input">
+					{ onSelectURL && <div className="block-editor-media-flow__url-input">
+						<span className="block-editor-media-replace-flow__image-url-label">{ __( ' Current media URL:' ) }</span>
```
```suggestion
						<span className="block-editor-media-replace-flow__image-url-label">{ __( 'Current media URL:' ) }</span>
```
I think we can remove the space.

shaunandrews commented:
I like it. Two suggestions:

1. Use a button with a label (instead of a mysterious icon) for the edit action. This matches up with the Link UI, and is also more obvious and clear to users.
2. Just general cleanup around the layout/display for editing the image URL. Here's a mockup of what I would expect it to look like:

<img width="627" alt="image" src="https://user-images.githubusercontent.com/191598/72278026-1b5a6e80-3601-11ea-809e-6b6e5e079789.png">

ItsJonQ commented:
@draganescu Hai there! Thank you for working on this fix 🙌 

I'm seeing some wonky UI from this PR. Just wanted to follow up to see if we were able to apply the changes suggested by @shaunandrews 

![Screen Capture on 2020-01-29 at 15-36-52](https://user-images.githubusercontent.com/2322354/73395705-0b53b780-42ae-11ea-9422-880bd3f4ecf2.gif)

Thanks again!

draganescu commented:
hi @ItsJonQ and thanks for testing!

Indeed, the `URLInput` component is raising some issues when used in this context. However now that the recent PRs that decoupled `LinkControl` from `Popover` are merged I will try to use that here in place of `URLInput`

draganescu commented:
This now works as expected but I am not happy with the implementation, as it still uses the old `LinkViewer` and `LinkEditor` components directly instead of the new `LinkControl`. However, because of how this is intended to work `LinkControl` is not ready yet because:

- it does not have a second "Apply" state
- cannot be closed

aduth commented:
Hi @draganescu , I'm not sure I follow what you mean by the specific requirements you list. What is the second "Apply" state, and how would a LinkControl be closed? Is this referring to the ability to have some option to "Cancel" out of the editing input and return to the preview state?

draganescu commented:
Yes, maybe I should have explained better. But briefly, if you look at the GIF up in the description of this PR, the interaction is in two steps:

1. you see the link
2. you edit the link

And this is very similar to how the `LinkControl` currently works, but except that in step 2, in this case we'd need to "Apply" the link edit and get back to step 1. Currently `LinkControl` doesn't allow that because it assumes you always want that suggestion drawer, so the input to edit a link does not have a button to "Apply" the edit.

Second, if I want to quit editig and get back to viewing the link, `LinkControl` does not allow this as it assumes that "quit" means unmounting it (via ESC or focusing somewhere else). In this context, we want the menu to remain open and the `LinkControl` to return to the view state.

getdave commented:
> Currently LinkControl doesn't allow that because it assumes you always want that suggestion drawer, so the input to edit a link does not have a button to "Apply" the edit.

I think we should add an explicit "Submit" (aka "Apply") UI. We could make this optional. It would sit next to the "Reset" button.

> Second, if I want to quit editig and get back to viewing the link, LinkControl does not allow this as it assumes that "quit" means unmounting it (via ESC or focusing somewhere else). In this context, we want the menu to remain open and the LinkControl to return to the view state.

In this scenario you have (for example) clicked "Edit" but then decided you want to go back and not edit? But you can't do that. Would the Apply button solve this issue?

I think both the above are solvable but should be in a separate PR.

aduth commented:
@draganescu Thanks for clarifying. These are things I'm very interested in as well.

> And this is very similar to how the `LinkControl` currently works, but except that in step 2, in this case we'd need to "Apply" the link edit and get back to step 1. Currently `LinkControl` doesn't allow that because it assumes you always want that suggestion drawer, so the input to edit a link does not have a button to "Apply" the edit.

If I understand this correctly: The applied editing does exist as handled as a form submission, but there's nothing in the UI to indicate this (or to click on for submitting). It relies on the <kbd>Enter</kbd> key press to submit/apply.

If my interpretation is correct, I think this can serve as a testament that #19971 would be a good revision to pursue to improve this clarity.

> Second, if I want to quit editig and get back to viewing the link, `LinkControl` does not allow this as it assumes that "quit" means unmounting it (via ESC or focusing somewhere else). In this context, we want the menu to remain open and the `LinkControl` to return to the view state.

There was some similar challenges for implementing this in the paragraph (https://github.com/WordPress/gutenberg/pull/19462#discussion_r373078084). You might be interested in the discussion in #20007.

It might be tricky for your specific use-case. My first thought was that "Escape" could toggle the internal state of `<LinkControl />` to "cancel" the edit, like one of the options you mention. However, I think it's also important that "Escape" would also close the currently visible popover. I'm not sure how best to deal with a scenario where both: (a) we have an editable state to cancel, and (b) that editable state is within a popover.

Using focus as a basis for the editing state _might_ work, but it needs to account for the additional action buttons shown alongside the input (currently "Reset", maybe in the future "Submit" and/or "Advanced Settings").

There could also be an explicit Cancel button in here as well. The problems here are that: (a) the space is limited ([related](https://github.com/WordPress/gutenberg/pull/19971#issuecomment-580868181)) and (b) the goals of #20007 might be such that we _don't want_ the preview state to be reachable at least in the context of paragraph link editing workflows.

aduth commented:
Thinking more about Cancel:

- Is it even necessary? If you're not going to submit your changes, then there's not a practical difference between "cancel" vs. simply leaving the input in its editable state, unless the Preview included some critical information.
    - To clarify, I don't feel strongly about this point, it's more a thought experiment.
- Canceling edit in response to `blur` of the input and its controls could be a reasonable way to approach this, at least as a compromise to support a use-case like the one needed here.

shaunandrews commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -15,15 +15,28 @@
 	padding-bottom: 10px;
+	border-top: 1px solid $light-gray-800;
+	margin-top: 5px;
+	padding-top: 15px;
```
There's three different padding attributes defined. I think we could reduce it down to just one. Also, maybe we can reduce the spacing a little — right now it feels a little _too_ big, especially with some of the other padding on child elements. I think updating to `padding: 12px 16px 8px 16px;` seems to work better visually when coupled with the other changes below.

shaunandrews commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -15,15 +15,28 @@
+}
+
 .block-editor-media-replace-flow__link-viewer {
+	margin-left: -3px;
```
We can get rid of this negative margin if we just override the margin on `.block-editor-url-popover__link-viewer-url` to `margin: 0;`.

shaunandrews commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -15,15 +15,28 @@
+
 .block-editor-media-replace-flow__link-viewer {
+	margin-left: -3px;
+
```
We can also add `align-items: center;` to better align the button and the link.

shaunandrews commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -15,15 +15,28 @@
+.block-editor-media-replace-flow__image-url-label {
+	color: $dark-gray-300;
+	font-size: $default-font-size;
+	padding-left: 4px;
```
No need for this extra padding here.

shaunandrews reviewed (commented):
Had some visual tweaks for the CSS to help line up elements, along with cleaning up margins and paddings. Here's how it looks with the suggested updates:

<img width="348" alt="image" src="https://user-images.githubusercontent.com/191598/73856809-050c9080-4804-11ea-90a7-5de6632c4bb8.png">

draganescu commented:
Hi @shaunandrews I've added all your suggestions and it looks pretty neat. Let me know of any other tweaks :)

aduth commented:
Per previous comments about LinkControl applicability here, not that as of #20052, at least one of the issues you mentioned should now be resolved (a "Submit" button is always shown in the editing state).

shaunandrews commented:
Pushed a quick change to add some more spacing around the Media Library and Upload buttons. Here's how it looks:

<img width="332" alt="image" src="https://user-images.githubusercontent.com/191598/74188434-65446d80-4c1c-11ea-81e5-e2cd5635cfca.png">

🚢

draganescu commented:
This is looking good now, but it does not use `LinkControl`, which is what I am working on :)

draganescu commented:
@getdave @aduth I have implemented this UI using the `LinkViewer` but I had to add some changes to the component to work as needed in this situation. Therefore:

- I added a `showSuggestions` prop to turn off suggestions showing (as in the case of media links we don't have media suggestions)
- I disabled `jsx-a11y/no-noninteractive-element-interactions` for the form in the SearchInput component b/c I couldn't stop the key presses from propagating otherwise

Do you think these are good modifications? Is the `LinkControl` the correct control as this is not really a link? Should I go back to using `LinkViewer` and `LinkEditor`?

aduth commented on packages/block-editor/src/components/link-control/search-input.js:
```diff
@@ -54,19 +58,22 @@ const LinkControlSearchInput = ( {
-					handleLinkControlOnKeyDown( event );
-				} }
-				onKeyPress={ handleLinkControlOnKeyPress }
+				onFocus={ noop }
```
Does this need to be assigned at all?

draganescu replied:
I was getting some console error because of it not being assigned. I couldnt repro so I took this out.

aduth commented on packages/block-editor/src/components/media-replace-flow/index.js:
```diff
@@ -171,21 +122,22 @@ const MediaReplaceFlow = ( {
+							</span>
+							<LinkControl
+								value={ { url: mediaURLValue } }
+								settings={ false }
```
It might work, but technically it's not of a valid type.

https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/components/link-control#settings

I think this should be more correct for what you're trying to do:

```suggestion
								settings={ [] }
```

aduth commented on packages/block-editor/src/components/media-replace-flow/index.js:
```diff
@@ -171,21 +122,22 @@ const MediaReplaceFlow = ( {
 						<div className="block-editor-media-flow__url-input">
-							{ urlInputUIContent }
+							<span className="block-editor-media-replace-flow__image-url-label">
+								{ __( ' Current media URL:' ) }
```
Not sure the leading space is intentional?

```suggestion
								{ __( 'Current media URL:' ) }
```

If you did need a bit more space, I would think style-applied padding would be more reliable.

draganescu replied:
Oh surely I didn't mean to use a space instead of padding :)))) good catch!

aduth commented on packages/block-editor/src/components/url-popover/link-viewer.js:
```diff
@@ -9,7 +9,6 @@ import classnames from 'classnames';
 import { __ } from '@wordpress/i18n';
 import { ExternalLink, Button } from '@wordpress/components';
 import { safeDecodeURI, filterURLForDisplay } from '@wordpress/url';
-import { pencil } from '@wordpress/icons';
```
Do you need the changes in this file anymore?

draganescu replied:
no.

aduth commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -95,6 +95,7 @@ function LinkControl( {
 	value,
 	settings,
 	onChange = noop,
+	showSuggestions = true,
```
Can we document this new prop, both in the associated JSDoc types definition and in the README.md?

draganescu replied:
done.

draganescu committed:
rebased and addressed feedback review

draganescu committed:
removed useless space in button label

draganescu committed:
use a correct type for empty settings

Co-Authored-By: Andrew Duthie <andrew@andrewduthie.com>

draganescu commented:
@aduth thanks for the review, I've addressed everything you pointed out. Given that design-wise we have a go, a final code review would help this land :) 
Thanks again!

draganescu commented:
@aduth thanks for the review, I've addressed everything you pointed out. Given that design-wise we have a go, a final code review would help this land :) 
Thanks again!

draganescu commented:
@aduth thanks for the review, I've addressed everything you pointed out. Given that design-wise we have a go, a final code review would help this land :) 
Thanks again!

draganescu committed:
fixes arrow keys navigation

draganescu commented on packages/block-editor/src/components/link-control/search-input.js:
```diff
@@ -4,13 +4,29 @@
+const handleLinkControlOnKeyPress = ( event ) => {
+	const { keyCode } = event;
+	event.stopPropagation();
+	if ( keyCode === ENTER ) {
```
Hi @Soean good catch, it was some copy pasta. I removed these functions altogether.

draganescu commented:
@Soean could you re-review and see if everything works as expected? :D Thank you 💯

Soean commented:
@draganescu 
If I use a different language with a longer button text like "Bearbeiten" in (Deutsch / German), the button is not visible on mobile.

English
<img width="377" alt="Bildschirmfoto 2020-02-21 um 11 31 46" src="https://user-images.githubusercontent.com/695201/75004467-c92d2880-549d-11ea-8932-37fbbbefc1ab.png">

Deutsch / German
<img width="377" alt="Bildschirmfoto 2020-02-21 um 11 30 12" src="https://user-images.githubusercontent.com/695201/75004399-9125e580-549d-11ea-8786-5a0b53745fd5.png">

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -534,7 +540,9 @@ function LinkControl( {
 					} }
-					renderSuggestions={ renderSearchResults }
+					renderSuggestions={
+						showSuggestions ? renderSearchResults : false
```
As `renderSuggestions` is returning React elements would it be better to return `null` here to indicate that we don't wish to render anything.

https://reactjs.org/docs/conditional-rendering.html#preventing-component-from-rendering

draganescu replied:
Yes, good catch :)

getdave commented on packages/block-editor/src/components/link-control/search-input.js:
```diff
@@ -44,7 +45,20 @@ const LinkControlSearchInput = ( {
+			onKeyPress={ ( event ) => {
+				const { keycode } = event;
+				if ( keycode === ENTER ) {
+					selectSuggestionOrCurrentInputValue();
```
This caught my eye. I'm sure this code got removed in earlier PRs on `LinkControl`. Does the action not get handled by the `onSubmit` handler?

If not then we could consider adding a comment here to indicate why we're having to double up here as it wasn't immediately apparent to me.

draganescu replied:
There is this comment in the URLInput component which made this neccessary:
```
// If the suggestions are not shown or loading, we shouldn't handle the arrow keys
// We shouldn't preventDefault to allow block arrow keys navigation
```
As in this context we explicitly disable suggestions then we need to also handle stopping key propagation to observe typing. Otherwise The pop up closes.

Would you think there is a better way around this?

draganescu replied:
I would have to also take care of the Escape button as now tests are failing b/c the link editor is not being dismissed on ESC anymore :( ugh.

getdave commented on packages/block-editor/src/components/media-replace-flow/index.js:
```diff
@@ -171,21 +122,22 @@ const MediaReplaceFlow = ( {
+							</span>
+							<LinkControl
+								value={ { url: mediaURLValue } }
+								settings={ [] }
```
Is this empty value required in order for `LinkControl` to function? If so then we need to raise an issue to make sure it isn't required.

draganescu replied:
Yes it is required. There is a default setting of showing `openInNewTab` option and in case of editing URLs you don't need that.

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -339,6 +341,10 @@ function LinkControl( {
 	const getSearchHandler = useCallback(
 		( val, args ) => {
+			if ( ! showSuggestions ) {
+				return Promise.resolve( [] );
```
I've thought a few times about making `URLInput` able to handle non-Promise values by using the `Promise.resolve()` wrapper around the `fetchSuggestions` prop. I wonder what you think to this?

draganescu replied:
I would think this is a good idea as I had a few tries until I realized what's what. This code above is not quire nice, resolving with an empty array, when returning false or the empty array would be much more readable in context.

getdave commented on packages/block-editor/src/components/link-control/search-input.js:
```diff
@@ -44,7 +45,20 @@ const LinkControlSearchInput = ( {
+		<form
+			onSubmit={ selectSuggestionOrCurrentInputValue }
+			onKeyDown={ ( event ) => {
+				event.stopPropagation();
```
I'm trying to work out why this is necessary. As this reintroduces code that was removed could we consider a comment here to understand why it's necessary?

draganescu replied:
See previous conversation.

getdave reviewed (commented):
Nice work on updating this to use `LinkControl`. It's great that we're stress testing this component. I hope it wasn't too painful!?

I noticed a few things that caught my eye. Let me know if anything doesn't make sense.

draganescu commented:
@Soean good catch. After rebasing the new G2 UI was added and now it is a bit worse :) The button is not cut, the whole link editor is having an invisible scrollbar. I wonder what is the best solution here. I would simply hide the link on mobile, to make sure edit in all languages definitely fits. Another option would be to wrap the button ... but that raises other issues.

draganescu committed:
adds sending null when suggestions are not needed as it is proper in react rendering

draganescu committed:
fixes the ESCAPE key to close the dropdown menu

Soean commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -5,17 +5,56 @@
+		display: block;
+		position: relative;
+		top: $grid-size-large;
+		padding-left: $grid-size-large;
```
```suggestion
		padding-left: $grid-unit-20;
```

Soean commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -5,17 +5,56 @@
+		}
+		.block-editor-link-control__search-item.is-current {
+			width: auto;
+			padding: $grid-size-large;
```
```suggestion
			padding: $grid-unit-20;
```

Soean commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -5,17 +5,56 @@
+		margin-top: -16px;
+		.block-editor-link-control__search-item-title {
+			max-width: 180px;
+			margin-top: $grid-size-large;
```
```suggestion
			margin-top: $grid-unit-20;
```

Soean commented on packages/block-editor/src/components/media-replace-flow/style.scss:
```diff
@@ -5,17 +5,56 @@
+		font-size: $default-font-size;
+		display: block;
+		position: relative;
+		top: $grid-size-large;
```
```suggestion
		top: $grid-unit-20;
```

Soean reviewed (approved):
Looks great now!
We need to change `$grid-size-large` to `$grid-unit-20`, see #20529.

draganescu committed:
Update packages/block-editor/src/components/media-replace-flow/style.scss

Co-Authored-By: Sören Wrede <soerenwrede@gmail.com>

draganescu committed:
Update packages/block-editor/src/components/media-replace-flow/style.scss

Co-Authored-By: Sören Wrede <soerenwrede@gmail.com>

draganescu committed:
Update packages/block-editor/src/components/media-replace-flow/style.scss

Co-Authored-By: Sören Wrede <soerenwrede@gmail.com>

draganescu committed:
Update packages/block-editor/src/components/media-replace-flow/style.scss

Co-Authored-By: Sören Wrede <soerenwrede@gmail.com>

draganescu committed:
removes deprecated mixing

draganescu committed:
moves keyboard capture outside of LinkControlto account only for it's usage in a dropwdown for MediaFlow

shaunandrews committed:
Overriding the width, removing the hr, and cleaning up spacing with the new grid variables.

shaunandrews committed:
Adjusting the spacing around the input for editing the link URL.

shaunandrews committed:
I'm not sure these are actually doing anything.

shaunandrews commented:
I pushed some CSS changes. Here's how its looking now:

![Replace Menu](https://user-images.githubusercontent.com/191598/76451818-6812b980-63a6-11ea-833c-f25f90347d08.gif)

I'd be nice if the popover was left-aligned with the toggle button in the toolbar, but that might be something that needs to be done on the `Dropdown` level?

draganescu commented:
Thank you @shaunandrews for the tidy up tweaks. I am so glad this is finally merged :)

noisysocks commented:
Are we able to truncate the URL in the middle so that it appears as `localhost…Climbing-Wall.jpg` instead of `localhost:8888/wp-content/…`? It's not useful unless you can see the file name of the image. (Sorry if this has already been raised!)

draganescu commented:
Good suggestion @noisysocks ! I'll open an issue with that in it.

paaljoachim commented:
What is the purpose of having the URL of the current media always showing in the media replace control?

draganescu commented:
Done here it is: https://github.com/WordPress/gutenberg/issues/21100
